core/src/main/java/jenkins/views/Header.java

```diff
@@ -0,0 +1,43 @@
+package jenkins.views;
+
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+
+/**
+ * Extension point that provides capabilities to render a specific header
+ * 
+ * @author Ildefonso Montero
+ * @see JenkinsHeader
+ */
+public abstract class Header implements ExtensionPoint {
+    
+    private static Logger LOGGER = Logger.getLogger(JenkinsHeader.class.getName());
+    
+    /**
+     * Checks if header is enabled.
+     * @return if header is enabled
+     */
+    public abstract boolean isHeaderEnabled();
+
+    @Restricted(NoExternalUse.class)
+    @CheckForNull
+    public static Header get() {
+        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).collect(Collectors.toList());
+        if (!headers.isEmpty()) {
+            if (headers.size() > 1) {
+                LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
```

Comment:
but the `JenkinsHeader` is always present and hard coded to be enabled so if you try and have another Header you will always have 2.

if your implementation / extension needs to also provide an ExtensionFilter to work (which it seems like it) then why fallback to the Jenkins header, this will completely and utterly spam the logs for every page load IIUC.
Or alternatively do not use an ExtensionFilter to remove the base JenkinsHeader and just use the Ordinal (and then you can just create and install one, and it will be some much simpler here - just return `ExtensionList.lookup(Header.class).get(0)`

Reply:
:thinking: how is this going to work when there exist two or more headers with the same `ordinal` value? (Consider two headers with ordinal as `Integer.MAX_VALUE`). I would prefer to give plugin developers the capability to filter the extension to be loaded to ensure its header is going to be the one active and rendered. Does it make sense?

Reply:
> ). I would prefer to give plugin developers the capability to filter the extension to be loaded to ensure its header is going to be the one active and rendered. Does it make sense?

not to me. :)

Because right now both plugin developers are *forced* to write an `ExtentionFilter`, they would likely just write one that removed all `Header` implementations that where not their own... thus you would even fallback to the `JenkinsHeader` that was itself filtered so not an extension (and you create a new one for each page view) and not even warn the user why as the headers extensionList would be empty.

If you did not force them to do that - you would likely not get into that situation to begin with., they would still be free to do that if they desire - but in all reality this would to me be better handled by warning the user in the UI to remove some plugin.

Reply:
Got it, IUCC then the main concern is not about making it feasible to use the `ExtensionFilter` at all, it is about current implementation is forcing developers to use it. Good point. Then I will just rely on the `ordinal` value as proposed. Updated in: https://github.com/jenkinsci/jenkins/pull/5909/commits/6fe15520b3b7fa327c0c7ba89cf3f67da5f46ce9